client/gutenberg/extensions/publicize/settings-button.jsx

```diff
@@ -25,15 +25,32 @@ import { Component } from '@wordpress/element';
  * Internal dependencies
  */
 import { __ } from 'gutenberg/extensions/presets/jetpack/utils/i18n';
+import { getSiteFragment } from 'lib/route/path';
 
 class PublicizeSettingsButton extends Component {
+	getButtonLink() {
+		const siteFragment = getSiteFragment( window.location.pathname );
+
+		// WP.com wp-admin will expose a window._currentSiteId with the current site ID
+		// and for Calypso we'll detect the site fragment from the current route.
+		//
+		// If running in WP.com wp-admin or in Calypso, we redirect to Calypso sharing settings.
+		if ( window._currentSiteId || siteFragment !== 'post.php' ) {
```

Comment:
Environment detection is really pretty ugly. We can't simply use our 'normal' Calypso selectors, such as `! isJetpackSite( state )`, can we? :slightly_frowning_face:

Reply:
With what we have right now, we can't - importing `connect` and all necessary we'll hit problems with `config` on some envs (wp-admin). Perhaps if we expose Calypso's state as a wp.data store..

Again, this is supposed to be a temporary solution, the ideal solution being new endpoints and a modal that keeps the user on the same page: #28387

Reply:
The moment we need it for another block, I think we should consider a better, broader solution instead. Until then, let's keep this, and I'm also happy to add a `@TODO` comment and a `NOTE: Please don't use it as a general purpose method for environment detection` comment there.

Reply:
Does this need also `post-new.php` check in addition to `post.php`?

Reply:
Good call - done in e44204d9b899e3242fa0abb6d94857ad266b431c

Reply:
This is one of the few times where I'd rather encapsulate the actual logic that determines if we're on a Jetpack or WordPress.com site early on, and make that available as a helper function to other people. 

Reasons:
- I fully expect this use case to arise elsewhere. We might fix this one by a new Publicize connections modal, but there are going to be other cases that will want different behavior on Jetpack vs. WordPress.com
- It's not necessarily going to be 'us' who's going to be in this situation next, and if there is no easily available helper, they'll either write their own, or copy this implementation.
- This shares some similarity with https://github.com/Automattic/wp-calypso/pull/28935. We might want to use some similar middleware/intercepting approach here.

That said, we should really get a fix for this in today. I'd prefer to maybe move at least [the check](https://github.com/Automattic/wp-calypso/pull/28949/files#diff-6fe10e108cbacbf401898d68d78257e6R39) to a helper function in the Jetpack preset's utils, but if that seems like too much work, lets's just add the TODO and merge it.

Reply:
Okay, you managed to convince me 😉
 
I've added a helper here: #28998 and updated this PR to use it.

I'd appreciate a review of both PRs, and if @lezama could test again, I'd truly appreciate it, thanks!